### PR TITLE
Setup Preregistration Crowdfunding

### PR DESCRIPTION
**Pull request by manveerxyz**
It's useful to go through this PR commit-by-commit.

## Move purchase views/serializers into folders
Self-explanatory, mainly for organizability.

## Setup new document type `PREREGISTRATION`
Self-explanatory.

## Implement Preregistration Fundraising
- New `Fundraise` model
  - Defaults `end_date` to 30 days from now
  - Uses `Escrow` to store RSC and handle payout
  - Uses `Purchase` to track contributions
- New `FundraiseViewSet`
  - Tried to ensure all edge-cases are accounted for (e.g. closed/expired fundraise, insufficient funds, etc.)
  - Triggers payout automatically once fundraise is fulfilled
  - Only allow moderators to create new fundraise using:
```bash
curl -X POST 'https://<BASE_URL>/api/fundraise/' -H 'Authorization: Token <TOKEN>' -H 'Content-Type: application/json' -d '{"goal_amount": <GOAL>, "goal_currency": "USD", "unified_document_id": <ID>, "recipient_user_id": <ID>}' | json_pp
```

## Add fundraise tests
Self explanatory, tried to handle main edge cases.

### CODE REVIEW

File: src/purchase/serializers/fundraise_serializer.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                total_amount=models.Sum(
+                    "amount_decimal"
+                )  # Assuming amount_decimal is a field in Purchase
+            )
+            .order_by("-total_amount")[:3]
+        )  # Selecting top 3
+
+        # Fetch user instances for top contributors
+        user_ids = [contributor["user_id"] for contributor in top_contributors]
+        top_3_users = User.objects.filter(id__in=user_ids)
```
Comment (yattias) [inline src/purchase/serializers/fundraise_serializer.py:84]: Nice fetching top 3 users as opposed to every user

File: src/purchase/tests/test_fundraise_view.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        self.assertEqual(response.status_code, 400)
+
+    def test_create_fundraise_not_moderator(self):
+        user = create_random_authenticated_user("fundraise_views")
+        self.client.force_authenticate(user)
+        response = self._create_fundraise(self.post.id, user=user)
+
+        self.assertEqual(response.status_code, 403)
+
+    # Contribution tests
```
Comment (yattias) [inline src/purchase/tests/test_fundraise_view.py:121]: Good tests here. Can we create a couple more tests:
- One that ensures the status of a fundraise changes to "CLOSED" 
- One that ensure user cannot contribute if fundraise is "CLOSED"

**Comment by yattias** [comment]
@manveerxyz Code looks good, just one comment. realized that some stuff has been moved around and github makes it seems like these are new files. I think I managed to look at all the actual code that was changes and feel good about it.

File: src/purchase/tests/test_fundraise_view.py (hunk 2 above)
Comment (manveerxyz) [reply src/purchase/tests/test_fundraise_view.py:123]: The fundraise_fulfilled test already checks that the status changes to `CLOSED`, but I added two more tests:
- Cannot contribute if fundraise is `CLOSED`
- Cannot contribute if fundraise is expired

Comment (yattias) [reply src/purchase/tests/test_fundraise_view.py:123]: Awesome. TY
